azkaban-common/src/main/java/azkaban/executor/container/watch/AzPodStatusDriver.java

```diff
@@ -0,0 +1,143 @@
+import io.kubernetes.client.util.Watch;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.TimeUnit;
+import java.util.function.Consumer;
+import javax.inject.Inject;
+import javax.inject.Singleton;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Drives execution {@link AzPodStatusListener} interface callbacks.
+ * Multiple listeners can be registered with a single instance of this class.
+ * Events received by this directly map the states in {@link AzPodStatus} and for every event all
+ * the callbacks registered for the particular event type are 'invoked'
+ *
+ * Invocation of the callbacks is done via pool of threads different from the one invoking this
+ * {@code onEvent} interface method of this class. This is done as a typical implementation of the
+ * {@link AzPodStatusListener} interface can include several blocking calls, for example for
+ * database status updates.
+ */
+@Singleton
+public class AzPodStatusDriver implements RawPodWatchEventListener {
+  private static final Logger logger = LoggerFactory.getLogger(AzPodStatusDriver.class);
+
+  private static final int THREAD_POOL_SIZE = 4;
+  public static final int SHUTDOW_TERMINATION_TIMEOUT_SECONDS = 5;
+  private final ExecutorService executor;
+  private final ImmutableMap<AzPodStatus, List<Consumer<AzPodStatusMetadata>>> listenerMap;
+
+  /**
+   * Create a new instance of {@link AzPodStatusDriver}.
+   */
+  @Inject
+  public AzPodStatusDriver() {
+    // Worth considering if we should get the thread-pool size from a config property instead.
+    executor = Executors.newFixedThreadPool(THREAD_POOL_SIZE,
+        new ThreadFactoryBuilder().setNameFormat("azk-watch-pool-%d").build());
+
+    ImmutableMap.Builder listenerMapBuilder = ImmutableMap.builder()
+        .put(AzPodStatus.AZ_POD_REQUESTED, new ArrayList<>())
+        .put(AzPodStatus.AZ_POD_SCHEDULED, new ArrayList<>())
+        .put(AzPodStatus.AZ_POD_INIT_CONTAINERS_RUNNING, new ArrayList<>())
+        .put(AzPodStatus.AZ_POD_APP_CONTAINERS_STARTING, new ArrayList<>())
+        .put(AzPodStatus.AZ_POD_READY, new ArrayList<>())
+        .put(AzPodStatus.AZ_POD_COMPLETED, new ArrayList<>())
+        .put(AzPodStatus.AZ_POD_INIT_FAILURE, new ArrayList<>())
+        .put(AzPodStatus.AZ_POD_APP_FAILURE, new ArrayList<>())
+        .put(AzPodStatus.AZ_POD_UNEXPECTED, new ArrayList<>());
+    listenerMap = listenerMapBuilder.build();
+  }
+
+  /**
+   * Register a new {@link AzPodStatusListener}.
+   * Note that the callbacks are submitted to the {@link ExecutorService} in the order interfaces
+   * were registered, but that does not make any guarantees about the sequence in which all the
+   * callbacks for a given event are invoked.
+   *
+   * @param listener
+   */
+  public void registerAzPodStatusListener(AzPodStatusListener listener) {
```

Comment:
using annotation @NonNull will be more helpful for input params as IntelliJ can help find a coding error beforehand compared to requireNoNull. Applies to any subsequent use for input params.

Reply:
I did not find a completion for @nonull, were you referring to `javax.annotation.Nonnull` perhaps ?
Good point, however I would prefer to retain the explicit call to `requireNonNull` as that allows for a custom exception message (and also conforms better to the rest of the code).

Reply:
Yes, the javax one is what I meant. The annotation is superior in this case. It finds the problem at compile time rather than execution time thus prevents bugs preemptively in most cases.
requireNonNull will find the bugs only after the code is deployed by throwing NPE.

Reply:
Dug a little more into it and it appears replacing `Objects.requireNonNull` with `javax.annotation.Nonnull` would be incorrect.
`requireNonNull` will force a runtime nullability check whereas the annotation `javax.annotation.Nonnull` will only introduce a compile time check that doesn't necessarily extend to runtime. 
An alternative for this is the annotation from project Lombok `lombok.NonNull` which does add a runtime nullability check as well: https://projectlombok.org/features/NonNull
However, I don't think introducing Lombok just for the nullability checks is a good idea. 
Additional information: https://stackoverflow.com/questions/48103792/what-is-the-difference-between-nonnull-and-objects-requirenonnull
Please let me know if you disagree.

Reply:
Thanks for digging this up. The same article says that they are rather complementary. I would recommend using both in this case to avoid programming errors.
Something, I also intend to do in my code in the future.

Reply:
The utility of the `javax.annotation.Nonnull` annotation (which will only do a static check) in addition to the `requireNonNull` check is a little questionable. 
Discussed offline and decided that introducing another `Nonnull` check needs more thought and should be done with a more comprehensive review of the available options and modifying the coding guidelines accordingly.